File: backend/modules/Portfolio/CromaticaCreativa.Modules.Portfolio.Domain/Entities/ProjectMedia.cs
using CromaticaCreativa.Modules.Portfolio.Domain.Enums;
using CromaticaCreativa.Modules.Portfolio.Domain.ValueObjects;

namespace CromaticaCreativa.Modules.Portfolio.Domain.Entities;

public sealed class ProjectMedia
{
    internal ProjectMedia(
        ProjectMediaId id,
        MediaReference reference,
        MediaType type,
        DisplayOrder order)
    {
        if (!Enum.IsDefined(type))
        {
            throw new ArgumentOutOfRangeException(nameof(type), type, "Unknown media type.");
        }

        Id = id ?? throw new ArgumentNullException(nameof(id));
        Reference = reference ?? throw new ArgumentNullException(nameof(reference));
        Type = type;
        Order = order;
    }

    public ProjectMediaId Id { get; }

    public MediaReference Reference { get; private set; }

    public MediaType Type { get; private set; }

    public DisplayOrder Order { get; private set; }

    internal void ChangeReference(MediaReference reference) =>
        Reference = reference ?? throw new ArgumentNullException(nameof(reference));

    internal void ChangeType(MediaType type)
    {
        if (!Enum.IsDefined(type))
        {
            throw new ArgumentOutOfRangeException(nameof(type), type, "Unknown media type.");
        }

        Type = type;
    }

    internal void ChangeOrder(DisplayOrder order) => Order = order;
}

File: backend/src/modules/Portfolio/Portfolio.Domain/Entities/ProjectMedia.ts
import { assertMediaType, MediaType } from '../Enums/MediaType';
import { DisplayOrder } from '../ValueObjects/DisplayOrder';
import { MediaReference } from '../ValueObjects/MediaReference';
import { ProjectMediaId } from '../ValueObjects/ProjectMediaId';

export class ProjectMedia {
  public constructor(
    public readonly id: ProjectMediaId,
    public readonly reference: MediaReference,
    public readonly type: MediaType,
    public readonly order: DisplayOrder,
  ) {
    assertMediaType(type);
  }

  public withChanges(reference: MediaReference, type: MediaType, order: DisplayOrder): ProjectMedia {
    return new ProjectMedia(this.id, reference, type, order);
  }

  public withOrder(order: DisplayOrder): ProjectMedia {
    return new ProjectMedia(this.id, this.reference, this.type, order);
  }
}
